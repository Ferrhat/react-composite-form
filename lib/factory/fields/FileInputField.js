import React, {Component} from "react";
import { Button } from 'react-bootstrap';
import { get } from 'lodash';

class FileInputField extends Component {

    render() {
        const { name, label, value, valueUpdateHandler, changeHandler, disabled } = this.props;
        let fileInput;
        let buttonLabel = label;
        let fileName = get(value, 'name');
        if (fileName) {
            buttonLabel += `: ${fileName}`;
        }


        const onChange = (event) => {
            valueUpdateHandler(name, event.target.files[0]);
            changeHandler(event);
        }

        const onClick = () => {
            if (fileInput) {
                fileInput.value = null;
                fileInput.click();
            }
        }

        return (
            <div>
                <Button onClick={onClick} disabled={disabled}>
                    <span className="glyphicon glyphicon-import" />
                    { buttonLabel }
                    <input type="file" style={{ display: 'none' }} onChange={onChange} ref={ref => fileInput = ref} />
                </Button>
                <div className={'validationErrorMessage'}>{get(this.props.errors, '0.message', '')}</div>
            </div>
        );
    }
}

export default FileInputField;
