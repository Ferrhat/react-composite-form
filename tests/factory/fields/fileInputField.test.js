import React from 'react';
import {mount} from 'enzyme';
import FileInputField from "../../../lib/factory/fields/FileInputField";


describe('FileInputField', () => {
    const mockValueUpdateHandler = jest.fn();
    const mockChangeHandler = jest.fn();
    const fileInputField = mount(<FileInputField valueUpdateHandler={mockValueUpdateHandler} changeHandler={mockChangeHandler} name={'testName'} />);

    it('renders properly', () => {
        expect(fileInputField).toMatchSnapshot();
    });

    it('contains an input component', () => {
        expect(fileInputField.find('Button').length).toEqual(1);
        fileInputField.find('Button').simulate('click');
        fileInputField.setProps({value: {name: 'testFileName'}});
        expect(fileInputField.find('input').length).toEqual(1);
    });

    it('should trigger a change event', () => {
        const event = {
            target: { files: ['test'] }
        };
        fileInputField.find('input').simulate('change', event);
        expect(mockValueUpdateHandler).toBeCalledWith('testName', 'test');
        expect(mockChangeHandler).toBeCalled();
    });

});
