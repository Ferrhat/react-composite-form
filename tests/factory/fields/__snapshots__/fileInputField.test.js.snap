// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInputField renders properly 1`] = `
<FileInputField
  changeHandler={[MockFunction]}
  name="testName"
  valueUpdateHandler={[MockFunction]}
>
  <div>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="glyphicon glyphicon-import"
        />
        <input
          onChange={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
        />
      </button>
    </Button>
    <div
      className="validationErrorMessage"
    />
  </div>
</FileInputField>
`;
